### PR TITLE
Fix blob DB transaction usage while GC

### PR DESCRIPTION
**Pull request by yiwu-arbug**
Summary:
While GC, blob DB use optimistic transaction to delete or replace the index entry in LSM, to guarantee correctness if there's a normal write writing to the same key. However, the previous implementation doesn't call SetSnapshot() nor use GetForUpdate() of transaction API, instead it do its own sequence number checking before beginning the transaction. A normal write can sneak in after the sequence number check and overwrite the key, and the GC will delete or relocate the old version of the key by mistake. Update the code to property use GetForUpdate() to check the existing index entry.

After the patch the sequence number store with each blob record is useless, So I'm considering remove the sequence number from blob record, in another patch.

Test Plan:
See the new tests. A similar test on existing implementation will fail.

### CODE REVIEW

**Comment by yiwu-arbug** [comment]
The internal phabricate diff will be a better place to view the diff.

**Comment by yiwu-arbug** [comment]
Synced with @siying yesterday and the new usage for optimistic transaction should be correct. @sagar0 mind review the rest of the logic in this patch? Thanks!

File: utilities/blob_db/blob_db_impl.cc [hunk 1]
```diff
@@ -1655,8 +1658,8 @@ std::pair<bool, int64_t> BlobDBImpl::WaStats(bool aborted) {
 // DELETED in the LSM
 ////////////////////////////////////////////////////////////////////////////////
 Status BlobDBImpl::GCFileAndUpdateLSM(const std::shared_ptr<BlobFile>& bfptr,
```
Comment (sagar0) [inline utilities/blob_db/blob_db_impl.cc]: 1. This function has grown too big. Lets work on splitting it up into smaller more manageable pieces, may be in a separate PR.
2. I also feel there are too many breaks and continues in the read-record loop, hindering the normal flow of understanding. It would be great if we can think of cleaning up that loop some more later.

Comment (yiwu-arbug) [inline utilities/blob_db/blob_db_impl.cc]: agree. Will do in a separate patch.
